File: Abdul-PB002/Wiki-Search/Abdul-PB002 (2).py
import requests
import sys
import json
import argparse
from datetime import datetime as dt

EXCEPTION_NONE = 0
EXCEPTION_VALUE_ERROR = 1

# Queries the API for a word and return the
# data and handles exception if something goes wrong
def getSearchResult(API_URL, query):
	payload = dict(
				action = 'opensearch',
				search = query,
				limit = '1',
				format = 'json'
				)
	try:
		result = requests.get(API_URL, params=payload)
		result = result.json()
		return EXCEPTION_NONE, result
	except ValueError as e:
		print "API returned no content\n"
		return EXCEPTION_VALUE_ERROR, None
	except requests.exceptions.RequestException as e:
		print "Something Went Wrong. Error Details"
		print e
		print "Exiting the program.\n"
		sys.exit(1)

# opens a file and returns a handler to it
def getLogFile(filename):
	logFile = None
	try:
		logFile = open(filename,'a')
	except IOError as e:
		print "Unable to open File\n"
		print e
		print "Cannot proceed. Terminating........."
		return None
	return logFile

# parse and save the data to the file
def saveLogtoFile(logFile, query, data, currentTime):
	try:
		title = data[1][0]
		url = data[3][0]
		dumpData = "{'Query': "+str(query)+", Time : "+ \
		currentTime+", 'Title' : "+str(title)+", 'URL' : "+str(url)+"})"+"\n"
		logFile.write(dumpData)
		print "Data saved in LogFile\n"
	except IOError as e:
		print "Unable to write to File\n"
		print e
	
	except ValueError as e:
		# print eof
		# print "Value Error"
		print "Wikipedia returned nothing\n"
	
	except IndexError as e:
		# print "Index Error"
		print "Wikipedia returned nothing\n"


def main():
	API_URL = "https://en.wikipedia.org/w/api.php"
	query = ""
	current_time = ""
	# Parser for the command line inputs # 
	parser = argparse.ArgumentParser(description=
		'Query a word to wikipedia and logs the links in a file')
	parser.add_argument('Filename', metavar='Filename', nargs=1,
						help='Log file')
	parser.add_argument('SearchQuery', metavar="QueryString", nargs='?',
						help='Search query string for wikipedia')
	args = parser.parse_args()
	####################################
	
	######### Opening Log File #########
	logFile = getLogFile(args.Filename[0])
	if not logFile:
		sys.exit(1)
	#########################################

	# Query Processing ######################
	

	if args.SearchQuery != None:	#If query is passed with command line
		print "You have passed "+args.SearchQuery + " as the search query\n"
		query = args.SearchQuery
	else:							#If query is not passed with command line	
		query = raw_input("Enter a search query string: ")
	
	#Runs untill user wants it to run
	while  query != "e" and query != "q" : 
		
		print "Fetching Data from Wikipedia..."
		current_time = str(dt.now()) # Recording the time of search
		returnCode, data = getSearchResult(API_URL,query)
		print "Done\n"


		if returnCode == EXCEPTION_NONE:	#If there is no error in API call
			saveLogtoFile(logFile, query, data,current_time)
		
		#Asks for another query or exit
		query = raw_input("Enter another query or type q or e to exit: ")

	# Gracefully closes the file
	try:
		logFile.close()
	except IOError as e:
		print e


if __name__ == '__main__':
	main()

File: Abdul-PB002/Wiki-Search/Abdul-PB002 (2).rb
require 'net/http'
require 'uri'
require 'json'



def getSearchResult(api_url,query)
	payload = { :action => 'opensearch', :search => query,
				:limit => 1, :format => 'json'}
	uri = URI(api_url)
	uri.query = URI.encode_www_form(payload)
	
	begin
		res = Net::HTTP.get_response(uri)
		
		jsonRes = JSON.parse(res.body)
		jsonRes
	rescue Timeout::Error, Errno::EINVAL, Errno::ECONNRESET, EOFError,
       Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, 
       Net::ProtocolError, SocketError => e
       puts e
       nil
   end
end





def saveLogToFile(logFile, query, data)
	title = data[1][0]
	url = data[3][0]
	dataDump = "{'Query' : "+query+", 'Title' : "+title+
				", 'URL' : "+url+"}"
	logFile.puts dataDump
end


api_url = 'https://en.wikipedia.org/w/api.php'
query = ''

# Handling command Line arguments
if(ARGV.length !=1 && ARGV.length !=2)
	puts "Usage : #{$0} filename [query]"
	exit
end


filename = ARGV[0]
logFile = File.open(filename,'a')
if(logFile == nil)
	puts "Unable to open a file"
	exit
end

if(ARGV[1]!=nil)
	query = ARGV[1]
else
	puts "Enter a query to search for: "
	query = STDIN.gets.to_s
	query = query.chomp
end

while query != "e" && query != "q" do
	puts "querying to Wikipedia...."
	result = getSearchResult(api_url,query)
	puts "Done\n"

	if (result == nil || result[3] == nil || result[3] == [])
		puts "Something Went wrong "
		puts "Terminating program.................."
		break
	else
		puts "Saving to log File"
		saveLogToFile(logFile, query, result)
		puts "Done\n"
		puts "Enter another query to search for or type q or e to exit: "
		query = STDIN.gets.to_s
		query = query.chomp
	end
end
logFile.close

